### PR TITLE
PM-4661: Add passkey.username as item.username

### PR DESCRIPTION
**Pull request by abergs**
## 🎟️ Tracking

https://bitwarden.atlassian.net/browse/PM-4661

## 📸 Screenshots

https://share.cleanshot.com/p1PH1krC

## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Protected functional changes with optionality (feature flags)
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation (Confluence, contributing docs) or informed the documentation team

## 🦮 Reviewer guidelines

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

### CODE REVIEW

File: libs/common/src/platform/services/fido2/fido2-authenticator.service.ts [hunk 1]
```diff
@@ -111,7 +111,7 @@ export class Fido2AuthenticatorService implements Fido2AuthenticatorServiceAbstr
       let pubKeyDer: ArrayBuffer;
       const response = await userInterfaceSession.confirmNewCredential({
         credentialName: params.rpEntity.name,
-        userName: params.userEntity.displayName,
+        userName: params.userEntity.name,
```
Comment (justindbaur) [inline libs/common/src/platform/services/fido2/fido2-authenticator.service.ts:114]: Is there any chance that `name` is `null` but `displayName` isn't so we'd want to do something like `params.userEntity.name ?? params.userEntity.displayName`?

**Review by gbubemismith — APPROVED** [review]
LGTM! Thanks for working on this @abergs

File: apps/browser/src/vault/popup/components/vault/add-edit.component.ts [hunk 2]
```diff
@@ -127,6 +127,10 @@ export class AddEditComponent extends BaseAddEditComponent {
 
       await this.load();
 
+      if (params.username) {
+        this.cipher.login.username = params.username;
+      }
```
Comment (justindbaur) [inline apps/browser/src/vault/popup/components/vault/add-edit.component.ts:132]: Could this be added in the below if block and guarded to only respect it if `this.cipher.login.username` is null or empty like the other ones? If the user has something saved it should always take precedence right? (I get that this param is brand new and only used for a brand new cipher but gotta defend against the future)

**Review by justindbaur — COMMENTED** [review]
Couple questions.

File: libs/common/src/platform/services/fido2/fido2-authenticator.service.ts (hunk 1 above)
Comment (abergs) [reply libs/common/src/platform/services/fido2/fido2-authenticator.service.ts:114]: Nope, name is required, but display name is not.

**Review by LRNcardozoWDF — DISMISSED** [review]
LGTM!

File: libs/common/src/platform/services/fido2/fido2-authenticator.service.ts (hunk 1 above)
Comment (abergs) [reply libs/common/src/platform/services/fido2/fido2-authenticator.service.ts:114]: right @coroiu?

File: apps/browser/src/vault/popup/components/vault/add-edit.component.ts (hunk 2 above)
Comment (abergs) [reply apps/browser/src/vault/popup/components/vault/add-edit.component.ts:132]: Was there any reason to resolving this @merissaacosta?

**Comment by abergs** [comment]
@kendratodd Perhaps you could help me test this unless it lands well into existing QA processes?

**Comment by kendratodd** [comment]
@abergs, I'm always happy to take a look, though I do believe this can be covered by existing QA processes for the Vault team (the JIRA just needs to be updated to reflect the current status).

File: libs/common/src/platform/services/fido2/fido2-authenticator.service.ts (hunk 1 above)
Comment (coroiu) [reply libs/common/src/platform/services/fido2/fido2-authenticator.service.ts:114]: Correct.

Technically both are `required`, but the spec allows setting the `displayName` to an empty string ([spec](https://www.w3.org/TR/webauthn-3/#dictionary-user-credential-params)), which is not allowed for `name` ([spec](https://www.w3.org/TR/webauthn-3/#dictionary-pkcredentialentity))

File: apps/browser/src/vault/popup/components/fido2/fido2.component.ts [hunk 3]
```diff
@@ -256,7 +256,7 @@ export class Fido2Component implements OnInit, OnDestroy {
       const name = data.credentialName || data.rpId;
       // TODO: Revert to check for user verification once user verification for passkeys is approved for production.
       // PM-4577 - https://github.com/bitwarden/clients/pull/8746
```
Comment (coroiu) [inline apps/browser/src/vault/popup/components/fido2/fido2.component.ts:258]: @gbubemismith should this comment be removed?

File: apps/browser/src/vault/popup/components/vault/add-edit.component.ts (hunk 2 above)
Comment (coroiu) [reply apps/browser/src/vault/popup/components/vault/add-edit.component.ts:132]: Unresolving, I agree that this should be addressed, we really shouldn't be overwriting usernames. I'm even hesitant to assign it even it was empty, since in some edge cases sites might not have a username

**Review by coroiu — CHANGES_REQUESTED** [review]
Tests failing and needs fix for comment from @justindbaur

File: apps/browser/src/vault/popup/components/vault/add-edit.component.ts [hunk 4]
```diff
@@ -128,6 +128,14 @@ export class AddEditComponent extends BaseAddEditComponent {
...
 
       if (!this.editMode || this.cloneMode) {
+        // Only allow setting username if there's no existing value
+        if (
+          params.username &&
+          (this.cipher.login.username === null || this.cipher.login.username === "")
+        ) {
+          this.cipher.login.username = params.username;
+        }
+
```
Comment (abergs) [inline apps/browser/src/vault/popup/components/vault/add-edit.component.ts:138]: @coroiu @justindbaur as per the feedback, is this ok?

Comment (coroiu) [reply apps/browser/src/vault/popup/components/vault/add-edit.component.ts:138]: You can also use `Utils.isNullOrWhitespace` or `Utils.isNullOrEmpty` if you want

Yes but I assume it would be assigned since the `Fido2Credential` is saved inside of the login?

Comment (abergs) [reply apps/browser/src/vault/popup/components/vault/add-edit.component.ts:138]: @coroiu So this would be Ok to approve?

Comment (coroiu) [reply apps/browser/src/vault/popup/components/vault/add-edit.component.ts:138]: I think so yes, but vault has the in-depth knowledge and needs to approve the changes :)

File: apps/browser/src/vault/popup/components/fido2/fido2.component.ts (hunk 3 above)
Comment (gbubemismith) [reply apps/browser/src/vault/popup/components/fido2/fido2.component.ts:258]: @coroiu, the comment is still needed. We just reverted from using user verification

File: apps/browser/src/vault/popup/components/vault/add-edit.component.ts (hunk 4 above)
Comment (coroiu) [reply apps/browser/src/vault/popup/components/vault/add-edit.component.ts:138]: I think so yes, but ultimately vault has the in-depth knowledge

Comment (gbubemismith) [reply apps/browser/src/vault/popup/components/vault/add-edit.component.ts:138]: @abergs this looks fine

File: apps/browser/src/vault/popup/components/vault/add-edit.component.ts [hunk 5]
```diff
@@ -128,6 +128,14 @@ export class AddEditComponent extends BaseAddEditComponent {
       await this.load();
 
       if (!this.editMode || this.cloneMode) {
+        // Only allow setting username if there's no existing value
+        if (
+          params.username &&
+          (this.cipher.login.username === null || this.cipher.login.username === "")
```
Comment (justindbaur) [inline apps/browser/src/vault/popup/components/vault/add-edit.component.ts:134]:
```suggestion
          (this.cipher.login.username == null || this.cipher.login.username === "")
```

Kind of a nit but also kind of not because JavaScript is a great language lol.

**Review by justindbaur — CHANGES_REQUESTED** [review]

**Review by justindbaur — APPROVED** [review]

**Review by coroiu — COMMENTED** [review]
Tests still failing

**Comment by abergs** [comment]
Let's see if I managed to fix the tests and actually run them now ...

image

**Review by coroiu — DISMISSED** [review]

**Review by gbubemismith — APPROVED** [review]

**Review by gbubemismith — APPROVED** [review]

**Review by coroiu — APPROVED** [review]
